tests/components/ipp/test_config_flow.py

```diff
@@ -0,0 +1,306 @@
+    assert result["description_placeholders"] == {CONF_NAME: "EPSON123456"}
+
+
+async def test_show_zeroconf_form(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test that the zeroconf confirmation form is served."""
+    aioclient_mock.post(
+        "http://EPSON123456.local:631/ipp/print",
+        content=load_fixture_binary("ipp/get-printer-attributes.bin"),
+        headers={"Content-Type": "application/ipp"},
+    )
+
+    flow = config_flow.IPPFlowHandler()
+    flow.hass = hass
+    flow.context = {"source": SOURCE_ZEROCONF}
+
+    discovery_info = MOCK_ZEROCONF_IPP_SERVICE_INFO.copy()
+    result = await flow.async_step_zeroconf(discovery_info)
+
+    assert flow.discovery_info[CONF_HOST] == "EPSON123456.local"
+    assert flow.discovery_info[CONF_NAME] == "EPSON123456"
+
+    assert result["step_id"] == "zeroconf_confirm"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["description_placeholders"] == {CONF_NAME: "EPSON123456"}
+
+
+async def test_connection_error(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test we show user form on IPP connection error."""
+    aioclient_mock.post(
+        "http://EPSON123456.local:631/ipp/print", exc=aiohttp.ClientError
+    )
+
+    user_input = MOCK_USER_INPUT.copy()
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": SOURCE_USER}, data=user_input,
+    )
+
+    assert result["step_id"] == "user"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["errors"] == {"base": "connection_error"}
+
+
+async def test_zeroconf_connection_error(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test we abort zeroconf flow on IPP connection error."""
+    aioclient_mock.post("http://EPSON123456.local/ipp/print", exc=aiohttp.ClientError)
+
+    discovery_info = MOCK_ZEROCONF_IPP_SERVICE_INFO.copy()
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": SOURCE_ZEROCONF}, data=discovery_info,
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "connection_error"
+
+
+async def test_zeroconf_confirm_connection_error(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test we abort zeroconf flow on IPP connection error."""
+    aioclient_mock.post("http://EPSON123456.local/ipp/print", exc=aiohttp.ClientError)
+
+    discovery_info = MOCK_ZEROCONF_IPP_SERVICE_INFO.copy()
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN,
+        context={
+            "source": SOURCE_ZEROCONF,
+            CONF_HOST: "EPSON123456.local",
+            CONF_NAME: "EPSON123456",
+        },
+        data=discovery_info,
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "connection_error"
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5274251)